# Week 4 Notes - Getting and Cleaning Data
Final week of the course! Let's get it! 

## Editing text variables
Imagine we have imported a data frame, and upon perusing it, we realise that the column names are formatted inconsistently, with variable cases, numbers and other characters in them, perhaps spaces too, making them more difficult to work with than they need to be - what can we do? 

In [2]:
using DataFrames, CSV, DataFramesMeta, HTTP, Dates

### mixed case formatting
In R we can use the **tolower()** function alongside the **names()** function to transform the column names to lower case - there is also likely an opposite **toupper()** function as well. 
```R
tolower(names(data))
```

In Julia we can directly rename the column names of the data frame to lowercase using the **rename()** function of DataFrames package
```julia
rename(lowercase, df)
```

And then some playing around with the functions directly

In [6]:
sample_names = ["MusicData", "tommorrowTemp", "AttRibutes"]

3-element Vector{String}:
 "MusicData"
 "tommorrowTemp"
 "AttRibutes"

In [7]:
lowercase.(sample_names)

3-element Vector{String}:
 "musicdata"
 "tommorrowtemp"
 "attributes"

### Stripping, splitting characters 

Say we have a variable which contains a dot followed by a number, and we're only interested in the string before the dot, we can strip and split these characters based on the character. In R;
```R
splitNames = strsplit(names(cameraData, "\\.")
```

In [17]:
sample_split = ["MusicData.3", "tommorrowTemp.3", "AttRibutes"]

3-element Vector{String}:
 "MusicData.3"
 "tommorrowTemp.3"
 "AttRibutes"

Replace all digits and numbers with nothing - effectively deleting it. 

In [45]:
replace.(sample_split, r"\d" => "", "." => "")

3-element Vector{String}:
 "MusicData"
 "tommorrowTemp"
 "AttRibutes"

This would be  done with something like **gsub** in R;
```R
gsub("_", "", testName)
```

Split the variable names on the dot character

In [46]:
split_names = split.(sample_split, ".")

3-element Vector{Vector{SubString{String}}}:
 ["MusicData", "3"]
 ["tommorrowTemp", "3"]
 ["AttRibutes"]

In [50]:
split_names[1][1]

"MusicData"

### Finding values - grep and so on
R borrows directly from the GNU grep function, whereas Julia uses it's own **find**, **findall** etc. functions for this purpose - along with **in**, **all** etc.. 

Let's see R's;
```R
grep("Alameda", data)
table(grepl("Alameda", data))
```

In julia we could use **findall()**;
```julia
findall("something", data)
```

## Dates and Time
The best guide on this in julia is found at https://en.wikibooks.org/wiki/Introducing_Julia/Working_with_dates_and_times -- very, very helpful 

## Quiz

## 1. 
The American Community Survey distributes downloadable data about United States communities. Download the 2006 microdata survey about housing for the state of Idaho using download.file() from here: 

https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2Fss06hid.csv

and load the data into R. The code book, describing the variable names is here:

https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FPUMSDataDict06.pdf 

Apply strsplit() to split all the names of the data frame on the characters "wgtp". What is the value of the 123 element of the resulting list? 

In [9]:
q1_csv = CSV.read(HTTP.get("https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2Fss06hid.csv").body, DataFrame)

6496×188 DataFrame
  Row │ RT       SERIALNO  DIVISION  PUMA   REGION  ST     ADJUST   WGTP   NP  ⋯
      │ String1  Int64     Int64     Int64  Int64   Int64  Int64    Int64  Int ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │ H             186         8    700       4     16  1015675     89      ⋯
    2 │ H             306         8    700       4     16  1015675    310
    3 │ H             395         8    100       4     16  1015675    106
    4 │ H             506         8    700       4     16  1015675    240
    5 │ H             835         8    800       4     16  1015675    118      ⋯
    6 │ H             989         8    700       4     16  1015675    115
    7 │ H            1861         8    700       4     16  1015675      0
    8 │ H            2120         8    200       4     16  1015675     35
    9 │ H            2278         8    400       4     16  1015675     47      ⋯
   10 │ H            2428         8    500       4     16  1015675     51
   11 │ H            2677         8    400       4     16  1015675    114
  ⋮   │    ⋮        ⋮         ⋮        ⋮      ⋮       ⋮       ⋮       ⋮      ⋮ ⋱
 6487 │ H         1356413         8    800       4     16  1015675     44
 6488 │ H         1356517         8    200       4     16  1015675      0      ⋯
 6489 │ H         1356950         8    500       4     16  1015675     71
 6490 │ H         1357182         8    700       4     16  1015675     23
 6491 │ H         1357195         8    300       4     16  1015675     19
 6492 │ H         1357238         8    300       4     16  1015675    108      ⋯
 6493 │ H         1357340         8    400       4     16  1015675     58
 6494 │ H         1357874         8    900       4     16  1015675     32
 6495 │ H         1357880         8    500       4     16  1015675    123
 6496 │ H         1358490         8    700       4     16  1015675    353      ⋯
                                               180 columns and 6475 rows omitted

In [7]:
colnames = names(q1_csv)

188-element Vector{String}:
 "RT"
 "SERIALNO"
 "DIVISION"
 "PUMA"
 "REGION"
 "ST"
 "ADJUST"
 "WGTP"
 "NP"
 "TYPE"
 "ACR"
 "AGS"
 "BDS"
 ⋮
 "wgtp69"
 "wgtp70"
 "wgtp71"
 "wgtp72"
 "wgtp73"
 "wgtp74"
 "wgtp75"
 "wgtp76"
 "wgtp77"
 "wgtp78"
 "wgtp79"
 "wgtp80"

In [15]:
split_colnames = split.(colnames, "wgtp")

188-element Vector{Vector{SubString{String}}}:
 ["RT"]
 ["SERIALNO"]
 ["DIVISION"]
 ["PUMA"]
 ["REGION"]
 ["ST"]
 ["ADJUST"]
 ["WGTP"]
 ["NP"]
 ["TYPE"]
 ["ACR"]
 ["AGS"]
 ["BDS"]
 ⋮
 ["", "69"]
 ["", "70"]
 ["", "71"]
 ["", "72"]
 ["", "73"]
 ["", "74"]
 ["", "75"]
 ["", "76"]
 ["", "77"]
 ["", "78"]
 ["", "79"]
 ["", "80"]

In [17]:
split_colnames[123]

2-element Vector{SubString{String}}:
 ""
 "15"

The answer is  "" "15"

## 2. 
Load the Gross Domestic Product data for the 190 ranked countries in this data set:

https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FGDP.csv

Remove the commas from the GDP numbers in millions of dollars and average them. What is the average?

Original data sources:

http://data.worldbank.org/data-catalog/GDP-ranking-table 

In [22]:
gdp = download("https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FGDP.csv", "gdp.csv") 

"gdp.csv"

In [23]:
run(`head gdp.csv`)

,Gross domestic product 2012,,,,,,,,
,,,,,,,,,
,,,,(millions of,,,,,
,Ranking,,Economy,US dollars),,,,,
,,,,,,,,,
USA,1,,United States," 16,244,600 ",,,,,
CHN,2,,China," 8,227,103 ",,,,,
JPN,3,,Japan," 5,959,718 ",,,,,
DEU,4,,Germany," 3,428,131 ",,,,,
FRA,5,,France," 2,612,878 ",,,,,


Process(`head gdp.csv`, ProcessExited(0))

In [31]:
gdp_raw = CSV.read("gdp.csv", DataFrame, skipto=6, header=4) ; gdp_raw_nomissing = dropmissing(gdp_raw[:, [:1, :2, :4, :5]]) ; gdp_cleaned = rename!(gdp_raw_nomissing, ["CountryCode", "Rank", "Country", "US_dollars"])

190×4 DataFrame
 Row │ CountryCode  Rank    Country                         US_dollars   
     │ String3      String  String31                        String15     
─────┼───────────────────────────────────────────────────────────────────
   1 │ USA          1       United States                    16,244,600
   2 │ CHN          2       China                            8,227,103
   3 │ JPN          3       Japan                            5,959,718
   4 │ DEU          4       Germany                          3,428,131
   5 │ FRA          5       France                           2,612,878
   6 │ GBR          6       United Kingdom                   2,471,784
   7 │ BRA          7       Brazil                           2,252,664
   8 │ RUS          8       Russian Federation               2,014,775
   9 │ ITA          9       Italy                            2,014,670
  10 │ IND          10      India                            1,841,710
  11 │ CAN          11      Canada                           1,821,424
  ⋮  │      ⋮         ⋮                   ⋮                      ⋮
 181 │ WSM          181     Samoa                            684
 182 │ COM          182     Comoros                          596
 183 │ DMA          183     Dominica                         480
 184 │ TON          184     Tonga                            472
 185 │ FSM          185     Micronesia, Fed. Sts.            326
 186 │ STP          186     S\xe3o Tom\xe9 and Principe      263
 187 │ PLW          187     Palau                            228
 188 │ MHL          188     Marshall Islands                 182
 189 │ KIR          189     Kiribati                         175
 190 │ TUV          190     Tuvalu                           40
                                                         169 rows omitted

In [32]:
gdp_cleaned.US_dollars .= replace.(gdp_cleaned.US_dollars, "," => "")

190-element Vector{String}:
 " 16244600 "
 " 8227103 "
 " 5959718 "
 " 3428131 "
 " 2612878 "
 " 2471784 "
 " 2252664 "
 " 2014775 "
 " 2014670 "
 " 1841710 "
 " 1821424 "
 " 1532408 "
 " 1322965 "
 ⋮
 " 767 "
 " 713 "
 " 684 "
 " 596 "
 " 480 "
 " 472 "
 " 326 "
 " 263 "
 " 228 "
 " 182 "
 " 175 "
 " 40 "

In [33]:
gdp_cleaned.US_dollars = parse.(Int, gdp_cleaned.US_dollars)

190-element Vector{Int64}:
 16244600
  8227103
  5959718
  3428131
  2612878
  2471784
  2252664
  2014775
  2014670
  1841710
  1821424
  1532408
  1322965
        ⋮
      767
      713
      684
      596
      480
      472
      326
      263
      228
      182
      175
       40

In [34]:
using StatsBase

In [35]:
mean(gdp_cleaned.US_dollars)

377652.4210526316

The answer is 377652.4210526316 

## 3. 
Question 3
In the data set from Question 2 what is a regular expression that would allow you to count the number of countries whose name begins with "United"? Assume that the variable with the country names in it is named countryNames. How many countries begin with United? 

In [61]:
contains.(gdp_cleaned.Country, "United")

190-element BitVector:
 1
 0
 0
 0
 0
 1
 0
 0
 0
 0
 0
 0
 0
 ⋮
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0

In [65]:
filter(x -> contains.(x.Country, "United"), gdp_cleaned)

3×4 DataFrame
 Row │ CountryCode  Rank    Country               US_dollars 
     │ String3      String  String31              Int64      
─────┼───────────────────────────────────────────────────────
   1 │ USA          1       United States           16244600
   2 │ GBR          6       United Kingdom           2471784
   3 │ ARE          32      United Arab Emirates      348595

In [42]:
occursin("United", "United States")

true

In [45]:
united_countries = occursin.("United", gdp_cleaned.Country)

190-element BitVector:
 1
 0
 0
 0
 0
 1
 0
 0
 0
 0
 0
 0
 0
 ⋮
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0
 0

In [46]:
countmap(united_countries)

Dict{Bool, Int64} with 2 entries:
  0 => 187
  1 => 3

## 4. 
Load the Gross Domestic Product data for the 190 ranked countries in this data set:

 https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FGDP.csv
 

Load the educational data from this data set:

https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FEDSTATS_Country.csv

Match the data based on the country shortcode. Of the countries for which the end of the fiscal year is available, how many end in June?

Same data as above, simply need to download the education data

In [66]:
download("https://d396qusza40orc.cloudfront.net/getdata%2Fdata%2FEDSTATS_Country.csv", "edu.csv")

"edu.csv"

In [68]:
edu_df = CSV.read("edu.csv", DataFrame)

234×31 DataFrame
 Row │ CountryCode  Long Name                          Income Group          R ⋯
     │ String3      String                             String31?             S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ ABW          Aruba                              High income: nonOECD  L ⋯
   2 │ ADO          Principality of Andorra            High income: nonOECD  E
   3 │ AFG          Islamic State of Afghanistan       Low income            S
   4 │ AGO          People's Republic of Angola        Lower middle income   S
   5 │ ALB          Republic of Albania                Upper middle income   E ⋯
   6 │ ARE          United Arab Emirates               High income: nonOECD  M
   7 │ ARG          Argentine Republic                 Upper middle income   L
   8 │ ARM          Republic of Armenia                Lower middle income   E
   9 │ ASM          American Samoa                     Upper middle income   E ⋯
  10 │ ATG          Antigua and Barbuda                Upper middle income   L
  11 │ AUS          Commonwealth of Australia          High income: OECD     E
  ⋮  │      ⋮                       ⋮                           ⋮              ⋱
 225 │ VNM          Socialist Republic of Vietnam      Lower middle income   E
 226 │ VUT          Republic of Vanuatu                Lower middle income   E ⋯
 227 │ WBG          West Bank and Gaza                 Lower middle income   M
 228 │ WLD          World                              missing               m
 229 │ WSM          Samoa                              Lower middle income   E
 230 │ YEM          Republic of Yemen                  Lower middle income   M ⋯
 231 │ ZAF          Republic of South Africa           Upper middle income   S
 232 │ ZAR          Democratic Republic of the Congo   Low income            S
 233 │ ZMB          Republic of Zambia                 Low income            S
 234 │ ZWE          Republic of Zimbabwe               Low income            S ⋯
                                                 28 columns and 213 rows omitted

In [70]:
edu_gdp = innerjoin(gdp_cleaned, edu_df, on="CountryCode")

189×34 DataFrame
 Row │ CountryCode  Rank    Country                         US_dollars  Long N ⋯
     │ String3      String  String31                        Int64       String ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ ABW          161     Aruba                                 2584  Aruba  ⋯
   2 │ AFG          105     Afghanistan                          20497  Islami
   3 │ AGO          60      Angola                              114147  People
   4 │ ALB          125     Albania                              12648  Republ
   5 │ ARE          32      United Arab Emirates                348595  United ⋯
   6 │ ARG          26      Argentina                           475502  Argent
   7 │ ARM          133     Armenia                               9951  Republ
   8 │ ATG          172     Antigua and Barbuda                   1134  Antigu
   9 │ AUS          12      Australia                          1532408  Common ⋯
  10 │ AUT          27      Austria                             394708  Republ
  11 │ AZE          68      Azerbaijan                           66605  Republ
  ⋮  │      ⋮         ⋮                   ⋮                     ⋮              ⋱
 180 │ VCT          180     St. Vincent and the Grenadines         713  St. Vi
 181 │ VEN          29      Venezuela, RB                       381286  Rep\xf ⋯
 182 │ VNM          57      Vietnam                             155820  Social
 183 │ VUT          177     Vanuatu                                787  Republ
 184 │ WSM          181     Samoa                                  684  Samoa
 185 │ YEM          90      Yemen, Rep.                          35646  Republ ⋯
 186 │ ZAF          28      South Africa                        384313  Republ
 187 │ ZAR          112     Congo, Dem. Rep.                     17204  Democr
 188 │ ZMB          104     Zambia                               20678  Republ
 189 │ ZWE          134     Zimbabwe                              9802  Republ ⋯
                                                 30 columns and 168 rows omitted

In [143]:
edu_gdp_special = filter(row -> any(!ismissing.(row.var"Special Notes")),edu_gdp)

66×34 DataFrame
 Row │ CountryCode  Rank    Country              US_dollars  Long Name         ⋯
     │ String3      String  String31             Int64       String            ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ AFG          105     Afghanistan               20497  Islamic State of  ⋯
   2 │ ATG          172     Antigua and Barbuda        1134  Antigua and Barbu
   3 │ AUS          12      Australia               1532408  Commonwealth of A
   4 │ AUT          27      Austria                  394708  Republic of Austr
   5 │ BEL          25      Belgium                  483262  Kingdom of Belgiu ⋯
   6 │ BGD          59      Bangladesh               116355  People's Republic
   7 │ BHS          138     Bahamas, The               8149  Commonwealth of T
   8 │ BLZ          169     Belize                     1493  Belize
   9 │ BMU          149     Bermuda                    5474  The Bermudas      ⋯
  10 │ BWA          117     Botswana                  14504  Republic of Botsw
  11 │ CAN          11      Canada                  1821424  Canada
  ⋮  │      ⋮         ⋮              ⋮               ⋮                      ⋮  ⋱
  57 │ SVN          80      Slovenia                  45279  Republic of Slove
  58 │ SWE          21      Sweden                   523806  Kingdom of Sweden ⋯
  59 │ SWZ          158     Swaziland                  3744  Kingdom of Swazil
  60 │ THA          31      Thailand                 365966  Kingdom of Thaila
  61 │ TKM          91      Turkmenistan              35164  Turkmenistan
  62 │ UGA          106     Uganda                    19881  Republic of Ugand ⋯
  63 │ URY          78      Uruguay                   49920  Oriental Republic
  64 │ USA          1       United States          16244600  United States of
  65 │ ZAF          28      South Africa             384313  Republic of South
  66 │ ZWE          134     Zimbabwe                   9802  Republic of Zimba ⋯
                                                  30 columns and 45 rows omitted

In [146]:
june_fiscal = filter(x -> contains.(x.var"Special Notes", "Fiscal year end: June"), edu_gdp_special)  

13×34 DataFrame
 Row │ CountryCode  Rank    Country           US_dollars  Long Name            ⋯
     │ String3      String  String31          Int64       String               ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ AUS          12      Australia            1532408  Commonwealth of Aust ⋯
   2 │ BGD          59      Bangladesh            116355  People's Republic of
   3 │ BWA          117     Botswana               14504  Republic of Botswana
   4 │ EGY          38      Egypt, Arab Rep.      262832  Arab Republic of Egy
   5 │ GMB          175     Gambia, The              917  Republic of The Gamb ⋯
   6 │ KEN          87      Kenya                  40697  Republic of Kenya
   7 │ KWT          56      Kuwait                160913  State of Kuwait
   8 │ PAK          44      Pakistan              225143  Islamic Republic of
   9 │ PRI          61      Puerto Rico           101496  Puerto Rico          ⋯
  10 │ SLE          157     Sierra Leone            3796  Republic of Sierra L
  11 │ SWE          21      Sweden                523806  Kingdom of Sweden
  12 │ UGA          106     Uganda                 19881  Republic of Uganda
  13 │ ZWE          134     Zimbabwe                9802  Republic of Zimbabwe ⋯
                                                              30 columns omitted

In [148]:
size(june_fiscal)

(13, 34)

Answer is 13 rows - 13 countries have a fiscal year which ends on June 30 - what a roller coaster!!!!! Patience and clear headedness are needed 

## 5. 
You can use the quantmod (
http://www.quantmod.com/
) package to get historical stock prices for publicly traded companies on the NASDAQ and NYSE. Use the following code to download data on Amazon's stock price and get the times the data was sampled. 